File: src/main/java/org/tindalos/principle/infrastructure/detector/submodulesblueprint/YAMLBasedSubmodulesBlueprintProvider.java
package org.tindalos.principle.infrastructure.detector.submodulesblueprint;

import java.io.File;
import java.io.IOException;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

import org.apache.commons.io.FileUtils;
import org.tindalos.principle.domain.core.ListConverter;
import org.tindalos.principle.domain.core.PackageReference;
import org.tindalos.principle.domain.detector.submodulesblueprint.InvalidBlueprintDefinitionException;
import org.tindalos.principle.domain.detector.submodulesblueprint.SubmoduleDefinition;
import org.tindalos.principle.domain.detector.submodulesblueprint.SubmoduleDefinitions;
import org.tindalos.principle.domain.detector.submodulesblueprint.SubmoduleDefinitionsProvider;
import org.tindalos.principle.domain.detector.submodulesblueprint.SubmoduleId;
import org.tindalos.principle.domain.expectations.SubmodulesDefinitionLocation;
import org.yaml.snakeyaml.Yaml;

import com.google.common.collect.Maps;
import com.google.common.collect.Sets;

public class YAMLBasedSubmodulesBlueprintProvider extends YAMLBasedSubmodulesBlueprintProviderScala {

	public SubmoduleDefinitions readSubmoduleDefinitions(SubmodulesDefinitionLocation submodulesDefinitionLocation, String basePackageName) {
		String yaml = getYAML(submodulesDefinitionLocation);
		return processYAML(yaml,basePackageName);
	}

	@SuppressWarnings("unchecked")
	private static SubmoduleDefinitions processYAML(String yamlText, String basePackageName) {
		
		Yaml yaml = new Yaml();
		Map<String, Object> yamlObject = (Map<String, Object>) yaml.load(yamlText);
		
		Map<SubmoduleId, SubmoduleDefinition> submoduleDefinitionMap = buildSubmoduleDefinitions(yamlObject, basePackageName);
		
		addDependencies(yamlObject, submoduleDefinitionMap);
		
		return new SubmoduleDefinitions(ListConverter.convert(submoduleDefinitionMap));
	
	}

	@SuppressWarnings("unchecked")
	private static void addDependencies(Map<String, Object> yamlObject, Map<SubmoduleId, SubmoduleDefinition> submoduleDefinitionMap) {
		
		Map<String, List<String>> dependencies = (Map<String,  List<String>>) yamlObject.get("subdmodule-dependencies");
		if (dependencies == null) {
			throw new InvalidBlueprintDefinitionException("Submodule dependencies not defined! ");
		}
		for (Entry<String,  List<String>> entry : dependencies.entrySet()) {
			SubmoduleId submoduleId = new SubmoduleId(entry.getKey());
			
			checkSubmoduleExists(submoduleDefinitionMap.keySet(), submoduleId);
			SubmoduleDefinition submoduleDefinition = submoduleDefinitionMap.get(submoduleId);
			
			Collection<SubmoduleId> plannedDependencies = transformToSubmoduleIds(entry.getValue(),submoduleDefinitionMap.keySet());
			submoduleDefinition.addPlannedDependencies(ListConverter.convert(plannedDependencies));
		}
	}

	@SuppressWarnings("unchecked")
	private static Map<SubmoduleId, SubmoduleDefinition> buildSubmoduleDefinitions(Map<String, Object> yamlObject, String basePackageName) {
		Map<SubmoduleId, SubmoduleDefinition> submoduleDefinitionMap = Maps.newHashMap();
		
		Map<String, List<String>> definitions = (Map<String,  List<String>>) yamlObject.get("subdmodule-definitions");
		if (definitions == null) {
			throw new InvalidBlueprintDefinitionException("Submodules not defined! ");
		}
		for (Entry<String,  List<String>> entry : definitions.entrySet()) {
			SubmoduleId submoduleId = new SubmoduleId(entry.getKey());
			
			Set<PackageReference> packages = transformToPackageReferences(entry.getValue(), basePackageName);
			SubmoduleDefinition submoduleDefinition = new SubmoduleDefinition(submoduleId, ListConverter.convert(packages));
			
			submoduleDefinitionMap.put(submoduleId, submoduleDefinition);
		}
		return submoduleDefinitionMap;
	}

	private static Collection<SubmoduleId> transformToSubmoduleIds(List<String> dependencies, Set<SubmoduleId> validSubmodules)  {
		
		Set<SubmoduleId> submodules = Sets.newHashSet();
		for (String submoduleName : dependencies) {
			SubmoduleId submoduleId = new SubmoduleId(submoduleName);
			checkSubmoduleExists(validSubmodules, submoduleId);
			submodules.add(submoduleId);
		}
		return submodules;
	}

	private static void checkSubmoduleExists(Set<SubmoduleId> validSubmodules, SubmoduleId submoduleId) {
		if (!validSubmodules.contains(submoduleId)) throw new InvalidBlueprintDefinitionException("No submodules defined with id " + submoduleId);
	}

	private static Set<PackageReference> transformToPackageReferences(List<String> packageNames, String basePackageName) {
		Set<PackageReference> packages = Sets.newHashSet();
		for(String packageName : packageNames) {
			packages.add(new PackageReference(basePackageName + "." + packageName));
		}
		return packages;
	}

	private static String getYAML(SubmodulesDefinitionLocation submodulesDefinitionLocation) {
		try {
			return FileUtils.readFileToString(new File(submodulesDefinitionLocation.filePath()));
		} catch (IOException e) {
			throw new InvalidBlueprintDefinitionException("problem with reading file from " + submodulesDefinitionLocation.filePath());
		}
	}

}

File: src/main/scala/org/tindalos/principle/infrastructure/detector/submodulesblueprint/YAMLBasedSubmodulesBlueprintProvider.scala
package org.tindalos.principle.infrastructure.detector.submodulesblueprint

import java.util

import org.tindalos.principle.domain.core.PackageReference
import org.tindalos.principle.domain.analyzers.submodulesblueprint.InvalidBlueprintDefinitionException
import org.tindalos.principle.domain.analyzers.submodulesblueprint.SubmoduleDefinition
import org.tindalos.principle.domain.analyzers.submodulesblueprint.SubmoduleId
import java.io.IOException
import org.apache.commons.io.FileUtils
import java.io.File
import org.yaml.snakeyaml.Yaml
import org.tindalos.principle.domain.analyzers.submodulesblueprint.SubmoduleDefinitions
import scala.collection.JavaConverters._

object YAMLBasedSubmodulesBlueprintProvider {

  def readSubmoduleDefinitions(submodulesDefinitionLocation: String, basePackageName: String) = {
    val yaml = getYAML(submodulesDefinitionLocation)
    processYAML(yaml, basePackageName)
  }

  protected def processYAML(yamlText: String, basePackageName: String) = {

    val yamlObject = new Yaml().load(yamlText).asInstanceOf[util.Map[String, Object]].asScala.toMap
    processYaml(basePackageName, yamlObject)
  }

  def processYaml(basePackageName: String, yamlObject: Map[String, Object]): SubmoduleDefinitions = {

    val checks: Map[String, Object] =
      yamlObject("checks")
          .asInstanceOf[java.util.Map[String, Object]]
          .asScala.toMap

    val modules: Map[String, Object] =
      checks("modules")
          .asInstanceOf[java.util.Map[String, Object]]
          .asScala.toMap

    val submoduleDefinitionMap = buildSubmoduleDefinitions(modules, basePackageName)
    addDependencies(modules, submoduleDefinitionMap)

    import scala.collection.JavaConverters._
    new SubmoduleDefinitions(submoduleDefinitionMap.asJava)
  }

  private def checkSubmoduleExists(validSubmodules: Set[SubmoduleId], submoduleId: SubmoduleId) =
    if (!validSubmodules.contains(submoduleId)) throw new InvalidBlueprintDefinitionException("No submodules defined with id " + submoduleId);

  private def transformToSubmoduleIds(dependencies: List[String], validSubmodules: Set[SubmoduleId]) = {

    val ids = dependencies.map(new SubmoduleId(_))
    ids.foreach(checkSubmoduleExists(validSubmodules, _))
    ids
  }

  private def addDependencies(yamlObject: Map[String, Object], submoduleDefinitionMap: Map[SubmoduleId, SubmoduleDefinition]) = {

    val dependenciesOpt = yamlObject.get("module-dependencies")
    if (dependenciesOpt.isEmpty) throw new InvalidBlueprintDefinitionException("Submodule dependencies not defined! ")

    val dependencies: Map[String, List[String]] = dependenciesOpt.map { d ⇒
      d.asInstanceOf[java.util.LinkedHashMap[String, java.util.List[String]]]
          .asScala.toMap
          .map { case (k, v) ⇒ (k, v.asScala.to[List]) }
    }.getOrElse(Map.empty)

    dependencies.foreach { keyVal =>
      val submoduleId = new SubmoduleId(keyVal._1)
      checkSubmoduleExists(submoduleDefinitionMap.keySet, submoduleId)
      val submoduleDefinition = submoduleDefinitionMap.get(submoduleId).get
      val plannedDependencies = transformToSubmoduleIds(keyVal._2, submoduleDefinitionMap.keySet)
      import scala.collection.JavaConverters._
      submoduleDefinition.addPlannedDependencies(plannedDependencies.asJava)
    }
  }

  private def buildSubmoduleDefinitions(yamlObject: Map[String, Object], basePackageName: String) = {

    val definitionsOpt = yamlObject.get("module-definitions")
    if (definitionsOpt.isEmpty) throw new InvalidBlueprintDefinitionException("Submodules not defined! ")

    val definitions: Map[String, List[String]] = definitionsOpt.map { ds ⇒
      ds.asInstanceOf[java.util.LinkedHashMap[String, java.util.List[String]]]
          .asScala.toMap
          .map { case (k, v) ⇒ (k, v.asScala.to[List]) }
    }.getOrElse(Map.empty)

    definitions.map { definition ⇒
      val submoduleId = new SubmoduleId(definition._1)
      val packages = transformToPackageReferences(definition._2, basePackageName)
      import scala.collection.JavaConverters._
      val javaPackages: java.util.Set[PackageReference] = packages.toSet.asJava
      val submoduleDefinition = new SubmoduleDefinition(submoduleId, javaPackages)
      (submoduleId, submoduleDefinition)
    }
  }

  private def transformToPackageReferences(packageNames: List[String], basePackageName: String) = packageNames.map(x => new PackageReference(basePackageName + "." + x))

  protected def getYAML(submodulesDefinitionLocation: String) =
    try {
      FileUtils.readFileToString(new File(submodulesDefinitionLocation))
    } catch {
      case ex: IOException => throw new InvalidBlueprintDefinitionException("problem with reading file from " + submodulesDefinitionLocation);
    }
}